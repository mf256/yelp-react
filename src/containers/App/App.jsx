import React from 'react';
import styled from 'styled-components';
import ApiError from '../../utils/ApiError';
import BusinessList from '../../components/BusinessList/BusinessList';
import GlobalStyles from '../../components/GlobalStyles/GlobalStyles';
import NavBar from '../../components/NavBar/NavBar';
import SearchBar from '../../components/SearchBar/SearchBar';
import WithLoading from '../../components/Loading/WithLoading';
import YelpApi from '../../utils/yelpApi';

const StyledApp = styled.div`
  max-width: 1260px;
  margin: 0 auto;
`;

class App extends React.Component {
  state = {
    loading: false,
    items: [],
    meta: {},
  };

  constructor(props) {
    super(props);
    this.searchYelp = this.searchYelp.bind(this);
    this.nextYelp = this.nextYelp.bind(this);
  }

  searchYelp(term, location, sortBy) {
    this.setState({
      loading: true,
    });

    try {
      YelpApi.search(term, location, sortBy).then((result) => {
        this.setState({
          loading: false,
          items: result.items,
          meta: result.meta,
        });
      });
    } catch (err) {
      if (err instanceof ApiError) {
        this.setState((state) => {
          const meta = Object.assign(state.meta, { error: err.message });
          return {
            loading: false,
            meta,
          };
        });
      }
    }
  }

  nextYelp() {
    const {
      items,
      meta: { term, location, sortBy },
    } = this.state;

    YelpApi.search(term, location, sortBy, items.length).then((result) => {
      this.setState((state) => {
        const newItems = state.items.concat(result.items);
        return {
          loading: false,
          items: newItems,
          meta: result.meta,
        };
      });
    });
  }

  render() {
    const {
      loading,
      items,
      meta: { error, total },
    } = this.state;
    const hasMore = items.length < total;
    const ListWithLoading = WithLoading(BusinessList);

    return (
      <StyledApp>
        <GlobalStyles />
        <NavBar />
        <SearchBar search={this.searchYelp} />
        <ListWithLoading
          businesses={items}
          isLoading={loading}
          errorMsg={error}
          loadMore={this.nextYelp}
          hasMore={hasMore}
        />
      </StyledApp>
    );
  }
}

export default App;
