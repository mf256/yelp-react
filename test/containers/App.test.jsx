import React from 'react';
import { shallow } from 'enzyme';
import App from '../../src/containers/App/App';

describe('App', () => {
  it('should match snapshot', () => {
    const component = shallow(<App />);

    expect(component).toMatchSnapshot();
  });
});
