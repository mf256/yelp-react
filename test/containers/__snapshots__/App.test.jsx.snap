// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match snapshot 1`] = `
<styled.div>
  <Memo(GlobalStyleComponent) />
  <NavBar />
  <SearchBar
    search={[Function]}
  />
  <WihLoadingComponent
    businesses={Array []}
    hasMore={false}
    isLoading={false}
    loadMore={[Function]}
  />
</styled.div>
`;
